core/shared/src/main/scala/zio/stm/TMap.scala

```diff
@@ -0,0 +1,185 @@
+    } yield res
+  }
+
+  /**
+   * Atomically performs side-effect for each binding present in map.
+   */
+  final def foreach[E](f: ((K, V)) => STM[E, Unit]): STM[E, Unit] =
+    self.foldM(())((_, kv) => f(kv))
+
+  /**
+   * Retrieves value associated with given key.
+   */
+  final def get[E](k: K): STM[E, Option[V]] =
+    for {
+      buckets <- tBuckets.get
+      idx     <- indexOf(k)
+      bucket  <- buckets(idx)
+    } yield bucket.find(_._1 == k).map(_._2)
+
+  /**
+   * Retrieves value associated with given key or default value, in case the
+   * key isn't present.
+   */
+  final def getOrElse[E](k: K, default: => V): STM[E, V] =
+    get(k).map(_.getOrElse(default))
+
+  /**
+   * Creates new [[TMap]] by mapping all bindings using pure function.
+   */
+  final def map[K2, V2](f: ((K, V)) => (K2, V2)): STM[Nothing, TMap[K2, V2]] =
+    self.fold(List.empty[(K2, V2)])((acc, kv) => f(kv) :: acc).flatMap(TMap(_))
+
+  /**
+   * Creates new [[TMap]] by mapping all bindings using effectful function.
+   */
+  final def mapM[E, K2, V2](f: ((K, V)) => STM[E, (K2, V2)]): STM[E, TMap[K2, V2]] =
+    self.foldM(List.empty[(K2, V2)])((acc, kv) => f(kv).map(_ :: acc)).flatMap(TMap(_))
+
+  /**
+   * Stores new binding into the map.
+   */
+  final def put[E](k: K, v: V): STM[E, Unit] = {
+    def upsert(bucket: List[(K, V)]): List[(K, V)] =
+      bucket match {
+        case Nil => List(k -> v)
+        case xs  => xs.map(kv => if (kv._1 == k) (k, v) else kv)
+      }
+
+    def resize(capacity: Int, data: List[(K, V)]): STM[Nothing, Unit] = {
+      val newCapacity = capacity * 2
+
+      for {
+        tmap       <- allocate(newCapacity, data)
+        newBuckets <- tmap.tBuckets.get
+        _          <- self.tBuckets.set(newBuckets)
+        _          <- self.tCapacity.set(newCapacity)
+      } yield ()
+    }
+
+    for {
+      buckets     <- tBuckets.get
+      idx         <- indexOf(k)
+      _           <- buckets.update(idx, upsert)
+      data        <- self.fold(List.empty[(K, V)])((acc, kv) => kv :: acc)
+      capacity    <- tCapacity.get
+      needsResize = capacity * LoadFactor < data.size
+      _           <- if (needsResize) resize(capacity, data) else STM.unit
+    } yield ()
+  }
+
+  /**
+   * Removes bindings matching predicate.
+   */
+  final def removeIf(p: ((K, V)) => Boolean): STM[Nothing, Unit] =
+    tBuckets.get.flatMap(_.transform(_.filterNot(p)))
+
+  /**
+   * Retains bindings matching predicate.
+   */
+  final def retainIf(p: ((K, V)) => Boolean): STM[Nothing, Unit] =
+    tBuckets.get.flatMap(_.transform(_.filter(p)))
+
+  private def indexOf(k: K): STM[Nothing, Int] =
+    tCapacity.get.map(c => k.hashCode() % c)
+}
+
+object TMap {
+
+  /**
+   * Makes a new `TMap` that is initialized with the specified values.
+   */
+  final def apply[K, V](data: => List[(K, V)]): STM[Nothing, TMap[K, V]] = {
+    val capacity = if (data.isEmpty) DefaultCapacity else 2 * data.size
```

Comment:
I would change to varargs (like most collections) and add a fromIterable method that takes an Iterable (not List).

Reply:
Done.